### ISSUE TITLE
[Feature]: React 19 support

### ISSUE THREAD

**Issue by eps1lon** [tremorlabs/tremor-npm#1072]
### What problem does this feature solve?

React 19 is already past its Alpha phase. During Alpha, libraries are encourage to start working on React 19 compatibility. This helps us flush out potential bugs before app developers start adopting React 19.

### What does the proposed API look like?

At a glance, `recharts` and `@headlessui/react` are the libraries that are incompatible that stand out.

`recharts@2.13.0-alpha.3` should fix most of the compat issues (already tracked in https://github.com/tremorlabs/tremor/issues/1054).

`@headlessui/react` had this compat issue: https://github.com/tailwindlabs/headlessui/issues/3167. https://github.com/tailwindlabs/headlessui/issues/3167 is already fixed on `main`. However, `@tremor/react` still uses 1.x so you'd need to bump `@headlessui/react` to 2.x first to have a chance of getting the [fix for `@headlessui/react`](https://github.com/tailwindlabs/headlessui/pull/3254).

**Comment by LUPLUV** [comment]
Will this be implemented soon? We would love to upgrade to react 19 while using tremor components :)

**Comment by michaelhays** [comment]
Upgrading [`package.json`](https://github.com/tremorlabs/tremor/blob/b1ac69817395397e81a7b636dd2958c31642609f/package.json#L33) to `recharts@2.13.0` should fix this: https://github.com/recharts/recharts/issues/4558

**Comment by rminami** [comment]
Currently getting this from `@headlessui/react` when trying to use Tremor with React v19.

```
./node_modules/@tremor/react/node_modules/@headlessui/react/dist/utils/stable-collection.js
Attempted import error: '__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED' is not exported from 'react' (imported as 'r').

Import trace for requested module:
./node_modules/@tremor/react/node_modules/@headlessui/react/dist/utils/stable-collection.js
./node_modules/@tremor/react/node_modules/@headlessui/react/dist/components/tabs/tabs.js
__barrel_optimize__?names=Tab!=!./node_modules/@tremor/react/node_modules/@headlessui/react/dist/headlessui.esm.js
./node_modules/@tremor/react/dist/components/input-elements/Tabs/TabGroup.js
__barrel_optimize__?names=Tab,TabGroup,TabList!=!./node_modules/@tremor/react/dist/index.js
```

**Comment by AaronMBMorse** [comment]
@rminami you can force the @headlessui/react to version 2.20 using resolutions in your package.json file. This isn't a great solution for a production application but might work as a holdover until it's updated.

**Comment by Elmosh** [comment]
Any updates on react 19 support?

**Comment by abdullahzen** [comment]
any updates on this?

**Comment by kevinr-security** [comment]
Are there any updates on this?

**Comment by vasiliyk** [comment]
hello 2026 :]

**Comment by vixalien** [comment]
it looks unmaintained

**Comment by babanomania** [comment]
Not official support, but if you're stuck waiting on this: I made a stopgap fork of [Tremor Raw](https://github.com/tremorlabs/tremor) (the Radix-based successor to this package, same team) with the dependencies updated. React 19.2, Tailwind v4, Recharts 3, react-day-picker 10, date-fns 4.

All 35+ components survived. Upstream's own Playwright suite (588 tests, three browsers) passes against the upgraded stack. You can install from a git tag or copy-paste components like before. The exact dependency changes are in [this commit](https://github.com/babanomania/tremor-react-19/commit/5e7eddfb4ab010591fb207876ba3f46d6f0c1da5), maybe useful as a reference if official support work picks up again.

```bash
npm install github:babanomania/tremor-react-19#v1.1.1
```

```css
/* your app CSS, that's the whole Tailwind setup */
@import "tailwindcss";
@import "tremor-react-19/theme.css";
```

Repo: https://github.com/babanomania/tremor-react-19, live Storybook: https://babanomania.github.io/tremor-react-19/

And an apology to the maintainers: I didn't do the upgrade by hand. I used Claude (Fable 5), which feels slightly wrong against work this carefully made. I made sure all the original tests still pass in the new repo before tagging anything.

Worth knowing before you switch: this follows Tremor Raw's API, not `@tremor/react`'s, so some components differ from what you're using here. Upstream was headed that way regardless. Once official React 19 support ships, use that instead, this fork is just a bridge.
